### PR TITLE
Servant.API.Modifiers

### PR DESCRIPTION
**Pull request by phadej**
Implements modifiers #856.

### Combinators

- [x] QueryParam
- [x] Header
- [x] ReqBody

`QueryParams`? `Required` -> `NonEmpty a`?

### Implementations

- [x] servant (Links)
- [x] servant-client-core
- [x] servant-server
- [x] servant-docs
- [x] servant-foreign
- [x] Amend ComprehensiveAPI

### QA

-  `Capture` can also be lenient, but I'll leave it for now, but by definition they are always required.

There's some info in the commit messages.

### Resolves

- https://github.com/haskell-servant/servant/issues/862
- When `ReqBody` is amended, will solve https://github.com/haskell-servant/servant/issues/793
- Supersedes https://github.com/haskell-servant/servant/pull/256

### CODE REVIEW

File: servant/servant.cabal [hunk 1]
```diff
@@ -49,12 +49,13 @@ library
...
     Servant.API.Internal.Test.ComprehensiveAPI
     Servant.API.IsSecure
+    Servant.API.Modifiers
     Servant.API.QueryParam
     Servant.API.Raw
-    Servant.API.Stream
     Servant.API.RemoteHost
     Servant.API.ReqBody
     Servant.API.ResponseHeaders
+    Servant.API.Stream
```
Comment (alpmestan) [inline servant/servant.cabal]: Oh, nice one. I'm feeling better already =P

File: servant/src/Servant/Utils/Links.hs [hunk 2]
```diff
@@ -101,8 +103,10 @@ module Servant.Utils.Links (
 import           Data.List
 import           Data.Monoid.Compat    ( (<>) )
 import           Data.Proxy            ( Proxy(..) )
+import           Data.Singletons.Bool  ( SBool (..), SBoolI (..) )
 import qualified Data.Text             as Text
 import qualified Data.Text.Encoding    as TE
+import           Data.Type.Bool (If)
```
Comment (jkarni) [inline servant/src/Servant/Utils/Links.hs]: Alignment

File: servant/src/Servant/Utils/Links.hs [hunk 3]
```diff
@@ -317,8 +330,33 @@ instance (ToHttpApiData v, HasLink sub)
...
+
+instance HasLink sub => HasLink (Summary s :> sub) where
+    type MkLink (Summary s :> sub) = MkLink sub
+    toLink _ = toLink (Proxy :: Proxy sub)
+
+instance HasLink sub => HasLink (HttpVersion :> sub) where
+    type MkLink (HttpVersion:> sub) = MkLink sub
+    toLink _ = toLink (Proxy :: Proxy sub)
+
+-- | /Note:/ the 'IsSecure' isn't visible in the resulting 'Link'.
```
Comment (jkarni) [inline servant/src/Servant/Utils/Links.hs]: I don't entirely understand this note.

File: servant/test/Servant/Utils/LinksSpec.hs [hunk 4]
```diff
@@ -75,11 +82,12 @@ spec = describe "Servant.Utils.Links" $ do
...
     it "can generate all links for an API that has only linkable endpoints" $ do
         let (allNames :<|> simple) = allLinks (Proxy :: Proxy LinkableApi)
-        simple
-          `shouldBeLink` "get"
-        allNames ["Seneca", "Aurelius"]
-          `shouldBeLink` "all/Seneca/Aurelius"
+        simple `shouldBeLink` "get"
+        allNames ["Seneca", "Aurelius"] `shouldBeLink` "all/Seneca/Aurelius"
 
+    it "can generate all inks for ComprehensiveAPIWithoutRaw" $ do
```
Comment (jkarni) [inline servant/test/Servant/Utils/LinksSpec.hs]: typo ("links")

File: servant/src/Servant/API/Modifiers.hs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+--
+-- >>> :kind! FoldRequired '[Required, Optional]
+-- FoldRequired '[Required, Optional] :: Bool
+-- = 'False
+--
+-- >>> :kind! FoldRequired '[]
+-- FoldRequired '[] :: Bool
+-- = 'False
+--
+type FoldRequired mods = FoldRequired' 'False mods
```
Comment (jkarni) [inline servant/src/Servant/API/Modifiers.hs]: Hmm, so here, another option is to have a map:

``` [IsRequired := 'True, Description := "something"] ```

(And optionally, `type Required = IsRequired := 'True ; type Optional = IsRequired := 'False`)

Which simplifies the implementation here, insofar as instead of having separate `FoldX`s, you can just have a generic "Lookup".

File: servant/src/Servant/API/Modifiers.hs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+-- | Fold a 'RequiredAgument' into a value
+foldRequiredArgument
+    :: forall mods a r. (SBoolI (FoldRequired mods))
+    => Proxy mods
+    -> (a -> r)        -- ^ 'Required'
+    -> (Maybe a -> r)  -- ^ 'Optional'
+    -> RequiredArgument mods a
+    -> r
+foldRequiredArgument _ f g mx =
+    case (sbool :: SBool (FoldRequired mods), mx) of
```
Comment (jkarni) [inline servant/src/Servant/API/Modifiers.hs]: Have you thought of making `SBool` an instance of `Reifies` (from `reflection`)? It seems like we might want non-boolean as well (e.g. `Description`), and then using a unified interface would be nice (rather than the one-off `SBoolI` class)

File: servant/src/Servant/API/Header.hs [hunk 7]
```diff
@@ -4,24 +4,25 @@
...
 --
 -- >>> newtype Referer = Referer Text deriving (Eq, Show)
 -- >>>
 -- >>>            -- GET /view-my-referer
 -- >>> type MyApi = "view-my-referer" :> Header "from" Referer :> Get '[JSON] Referer
-data Header (sym :: Symbol) a = Header a
-                              | MissingHeader
-                              | UndecodableHeader ByteString
-    deriving (Typeable, Eq, Show, Functor)
+type Header = Header' '[]
```
Comment (jkarni) [inline servant/src/Servant/API/Header.hs]: For these, should we maybe be explicit about `Optional` etc? That might serve as a little bit of documentation.

Comment (alpmestan) [reply servant/src/Servant/API/Header.hs]: +1 on that.

File: servant/src/Servant/API/Modifiers.hs (hunk 5 above)
Comment (phadej) [reply servant/src/Servant/API/Modifiers.hs]: Using map-like is an option.

TL;DR, current approach:
- pros: there isn't ill-kinded situations like `IsRequired := "true"`
- cons: slight boilerplate in the implementation (not use!)

IMHO pros outweight cons here.

---

I personally dislike "dynamic" feeling of `(:=) :: * -> k -> *`. Nothing really prevents you from saying `'[IsRequired := "true"]`. OTOH, having *typed* labels is a great improvements over `(:=) :: Symbol -> k -> *`, proposed earlier, as then typos in label name would get unnoticed.

Think about term level, we could have two APIs:

```haskell
defaultOptions &
  & required .~ True
  & description ?~ "Foo"
-- or
required <> description "Foo"
```

The type-level list approach is like the `Monoid` one.
Having all "type-safety" of lens one (value type is dependent on the key!) is tricky.
My gut feeling that in GHC `>= 8.0` one could encode that, but IMHO it's not worth that.

File: servant/src/Servant/API/Modifiers.hs (hunk 6 above)
Comment (phadej) [reply servant/src/Servant/API/Modifiers.hs]: `reflection` isn't powerful enough. We need something to pattern match on (`SingI` from `singletons`, which I'd like to avoid for now as a dependency) to force `RequiredArgument mods a` into either `a` or `Maybe a`.

Note: the instance would be
```haskell
instance SBoolI b => Reifies Bool Bool where
    reflect _ = case sbool :: ... of STrue -> True; SFalse -> False
```

File: servant/src/Servant/Utils/Links.hs (hunk 3 above)
Comment (phadej) [reply servant/src/Servant/Utils/Links.hs]: `linkURI :: Link -> URI` should reflect the scheme, we use `IsSecure` combinator for links too, `f IsSecure = "https"; f NotSecure = "http"`.

**Comment by alpmestan** [comment]
@taktoa I believe you were asking about this on #nixos-dev yesterday. Didn't find you on IRC so I'm pinging you here: this is the upcoming solution to the problem you were asking about.

**Comment by taktoa** [comment]
@alpmestan Yup, I actually linked this in there. I was mainly asking because I thought someone might have an out-of-tree version of the code that I could use more easily than this patch, but I realized that I don't have a near-term need for the `HasServer` functionality, so I can just use the version that only has `HasClient` until this gets merged and upstreamed to Hackage.

File: servant-foreign/test/Servant/ForeignSpec.hs [hunk 8]
```diff
@@ -7,9 +7,11 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE PolyKinds #-}
 {-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE TemplateHaskell #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
+#if __GLASGOW__HASKELL < 709
+{-# OPTIONS_GHC -fcontext-stack=41 #-}
```
Comment (alpmestan) [inline servant-foreign/test/Servant/ForeignSpec.hs]: :+1:

**Comment by phadej** [comment]
Green. I'm positively surprised that `servant-js` isn't broken. I'll reread what I wrote tomorrow, and then merge.
